include/particles/twophase_particle.tcc

```diff
@@ -0,0 +1,735 @@
+                               liquid_mass_ * shapefn_[i] * liquid_velocity_);
+  }
+}
+
+//! Compute pore pressure
+template <unsigned Tdim>
+void mpm::TwoPhaseParticle<Tdim>::compute_pore_pressure(double dt) noexcept {
+  // Check if liquid material and cell pointer are set and positive
+  assert(this->material(mpm::ParticlePhase::Liquid) != nullptr &&
+         cell_ != nullptr);
+
+  // get the bulk modulus of liquid
+  double K = this->material(mpm::ParticlePhase::Liquid)
+                 ->template property<double>(std::string("bulk_modulus"));
+
+  // Compute at centroid
+  // get liquid phase strain rate at cell centre
+  auto liquid_strain_rate_centroid =
+      this->compute_strain_rate(dn_dx_centroid_, mpm::ParticlePhase::Liquid);
+
+  // update pressure
+  this->state_variables_[mpm::ParticlePhase::Liquid].at("pressure") +=
+      -dt * (K / porosity_) *
+      ((1 - porosity_) * strain_rate_.head(Tdim).sum() +
+       porosity_ * liquid_strain_rate_centroid.head(Tdim).sum());
+
+  // Apply free surface
+  if (this->free_surface())
+    this->assign_state_variable("pressure", 0., mpm::ParticlePhase::Liquid);
+}
+
+//! Map body force for both mixture and liquid
+template <unsigned Tdim>
+void mpm::TwoPhaseParticle<Tdim>::map_body_force(
+    const VectorDim& pgravity) noexcept {
+  this->map_mixture_body_force(mpm::ParticlePhase::Mixture, pgravity);
+  this->map_liquid_body_force(pgravity);
+}
+
+//! Map liquid phase body force
+template <unsigned Tdim>
+void mpm::TwoPhaseParticle<Tdim>::map_liquid_body_force(
+    const VectorDim& pgravity) noexcept {
+  // Compute nodal liquid body forces
+  for (unsigned i = 0; i < nodes_.size(); ++i)
+    nodes_[i]->update_external_force(
+        true, mpm::ParticlePhase::Liquid,
+        (pgravity * this->liquid_mass_ * shapefn_(i)));
+}
+
+//! Map mixture body force
+template <unsigned Tdim>
+void mpm::TwoPhaseParticle<Tdim>::map_mixture_body_force(
+    unsigned mixture, const VectorDim& pgravity) noexcept {
+  // Compute nodal mixture body forces
+  for (unsigned i = 0; i < nodes_.size(); ++i)
+    nodes_[i]->update_external_force(
+        true, mixture,
+        (pgravity * (this->liquid_mass_ + this->mass_) * shapefn_(i)));
+}
+
+//! Map traction force
+template <unsigned Tdim>
+void mpm::TwoPhaseParticle<Tdim>::map_traction_force() noexcept {
+  this->map_mixture_traction_force();
+  this->map_liquid_traction_force();
+}
+
+//! Map mixture traction force
+template <unsigned Tdim>
+void mpm::TwoPhaseParticle<Tdim>::map_mixture_traction_force() noexcept {
+  if (this->set_traction_) {
+    // Map particle traction forces to nodes
+    for (unsigned i = 0; i < nodes_.size(); ++i)
+      nodes_[i]->update_external_force(true, mpm::ParticlePhase::Mixture,
+                                       (shapefn_[i] * traction_));
+  }
+}
+
+//! Map liquid traction force
+template <unsigned Tdim>
+void mpm::TwoPhaseParticle<Tdim>::map_liquid_traction_force() noexcept {
```

Comment:
Shall we activate this?

Reply:
Please check the new design for `liquid_traction`. I use the original `assign_traction` function, and 0~Tdim-1 for mixture, Tdim~2Tdim-1 for liquid. @bodhinandach